autoreject/autoreject.py

```diff
@@ -594,46 +641,23 @@ def transform(self, epochs):
         """
         _check_data(epochs, picks=self.picks, verbose=self.verbose,
                     ch_constraint='data_channels')
-        if not all(epochs.ch_names[pp] in self.threshes_ for pp in self.picks):
+        if not all(epochs.ch_names[pp]
+                   in self.threshes_ for pp in self.picks_):
             raise ValueError('You are passing channels which were not present '
                              'at fit-time. Please fit it again, this time '
                              'correctly.')
         epochs_out = epochs.copy()
-        sub_picks = _check_sub_picks(picks=self.picks, info=epochs_out.info)
-        if sub_picks is not False:
-            bad_epochs_idx = list()
-            interp_channels_list = list()
-            old_picks = self.picks
-            for ii, (ch_type, this_picks) in enumerate(sub_picks):
-                self.picks = this_picks
-                out = self._annotate_epochs(self.threshes_, epochs)
-                interp_channels_list.append(out[2])
-                bad_epochs_idx_ = out[4]
-                bad_epochs_idx = np.union1d(bad_epochs_idx, bad_epochs_idx_)
-            good_epochs_idx = np.setdiff1d(np.arange(len(epochs)),
-                                           bad_epochs_idx)
-            if len(good_epochs_idx) == 0:
-                raise ValueError('All epochs are bad. Sorry.')
-
-            for ii, (ch_type, this_picks) in enumerate(sub_picks):
-                self.picks = this_picks
-                self._interpolate_bad_epochs(
-                    epochs_out, interp_channels=interp_channels_list[ii],
-                    verbose=self.verbose)
-            self.picks = old_picks
-
-        else:
-            (_, _, interp_channels, _,
-             bad_epochs_idx, good_epochs_idx) = self._annotate_epochs(
-                 threshes=self.threshes_, epochs=epochs)
-            if len(good_epochs_idx) == 0:
-                raise ValueError('All epochs are bad. Sorry.')
+        annot = self.annotate_epochs(epochs_out, picks=None)
+        if len(annot['good_epochs_idx']) == 0:
+            raise ValueError('All epochs are bad. Sorry.')
 
+        for ch_type, this_picks in annot['picks_by_type']:
             self._interpolate_bad_epochs(
-                epochs_out, interp_channels=interp_channels,
-                verbose=self.verbose)
-        if np.any(bad_epochs_idx):
-            epochs_out.drop(bad_epochs_idx, reason='AUTOREJECT')
+                epochs_out, interp_channels=annot['interp_channels'][ch_type],
+                picks=this_picks, verbose=self.verbose)
+
+        if np.any(annot['bad_epochs_idx']):
+            epochs_out.drop(annot['bad_epochs_idx'], reason='AUTOREJECT')
```

Comment:
I'm wondering if this should be inside annotate_epochs and we do away with good_epochs_idx and bad_epochs_idx, at least in the public API?